doc: add unhandled promise rejection survey

mmarchini opened the pull request:
Since we decided to have a survey before voting on https://github.com/nodejs/node/pull/33021, I started to draft the survey to capture community feedback. This is by no means the final version of the survey, it's just a starting point as I believe we should build this survey together.

Please leave comments suggestion questions we should include or remove, or any changes to the existing questions from title to description to available options.

ljharb commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,68 @@
+ - [ ] Both
+ - [ ] Neither / not writing Promise-based code
+
+## Are you currently using `.catch()`, `try await catch`, a mix, or neither to handle rejections?
```
"`try await catch`" isn't an idiom i've heard to describe what you're describing here; i think it needs elaboration.

ljharb commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,68 @@
+
+## Do you use one or more of the following options as a global handler for Promise rejections?
+
+Leave empty if you use the default Node.js behavior
```
maybe better to have an explicit option for "use the default behavior", otherwise you can't necessarily infer it based on "didn't answer this question"

jasnell replied:
Absolutely. There are many who do not really have an understanding of what the default Node.js behavior actually is, even if they do use it. Alternatively, I would split this into two separate yes/no questions:

* Do you use `process.on('unhandledRejection')` as a global handler for Promise rejections?
* Do you use the `--unhandled-rejections` command-line flag as a global handler for Promise rejections?

Ginden commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,68 @@
+  - [ ] Third-party libraries
+  - [ ] Command-line tools
+  - [ ] Web servers
+  - [ ] Computational-heave processing
```
typo: should be "heavy"

Trott replied:
```suggestion
  - [ ] Computation-heavy processing
```

puzpuzpuz commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,68 @@
+
+Leave empty if you use the default Node.js behavior
+
+  - [ ] Yes, via process.on('unhandledRejection')
```
```suggestion
  - [ ] Yes, via `process.on('unhandledRejection')`
```

jasnell commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,68 @@
+ - [ ] `.catch()`
+ - [ ] `try await catch`
+ - [ ] Both
+ - [ ] Neither / not writing Promise-based code
```
I would add an other option here. Some codebases rely on a mixture of strategies and it's often inconsistent when there are multiple teams involved so this can be a difficult question to ask.

`Ignore the rejection` should likely also be an explicit option here.

jasnell commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,68 @@
+
+## When a rejected promise doesn't have a catch handler, what is the behavior of your application today? 
+
+  - [ ] Current default Node.js behavior: logs a warning alongside a deprecation notice, continue execution
```
```suggestion
  - [ ] Node.js logs a warning alongside a deprecation notice, execution continues
```

mcollina commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,68 @@
+ - [ ] `try await catch`
+ - [ ] Both
+ - [ ] Neither / not writing Promise-based code
+
```
I would add the following question:

```
Are you aware that not handling rejections may cause memory leaks or denial of service vulnerabilities?

- [  ] Yes
- [  ] No
- [  ] unhandled rejections do not cause memory leaks or denial of service vulnerabilities
```

mmarchini commented:
I think I addressed all comments. I turned the first two questions into multiple-choice questions (still requiring choosing at least one answer), broke the global handler question into two Yes/No questions, and added the memory leak question as the last one. Let me know if y'all have any other suggestions.

ljharb commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,80 @@
+
+## Are you currently using Promises, async/await, a mix, or neither?
+
+Assume using as creating your own `new Promise` and `async function`
```
```suggestion
Assume using as creating your own `new Promise`, `Promise.resolve`, and `async function`; or using libraries that produce Promises (that you use with `.then`/`.catch`, or with `.await`)
```

ljharb commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,80 @@
+
+  - [ ] Promises
+  - [ ] async/await
+  - [ ] Not writing Promise-based code
```
```suggestion
  - [ ] Not writing Promise-based code or using Promise-based libraries
```

ljharb commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,80 @@
+  - [ ] `.catch()`
+  - [ ] `try / catch` wrapping an `await` operation
+  - [ ] Ignore the rejection
+  - [ ] Not writing Promise-based code
```
```suggestion
  - [ ] Not writing Promise-based code or using Promise-based libraries
```

ljharb commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,80 @@
+  - [ ] No
+  - [ ] I don't know
+
+## Which of the use cases described below do you spend most of your time on?
```
this should probably be multiple choice?

mmarchini replied:
Might make sense, although I was trying to capture majority of time spent (and not everything someone works on).

devsnek commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,80 @@
+  - [ ] `none`
+  - [ ] Other (please elaborate)
+
+# Are you aware that not handling rejections may cause memory leaks or denial of service vulnerabilities?
```
This seems somewhat leading/argumentative (with the third answer below). What is the goal of this question?

mmarchini replied:
@mcollina mind to shed some light on the goal of this question?

benjamingr replied:
I am also not a fan of this question and think it's leading the participants. I would ask about the actual objective issue that is being discussed.

ljharb commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,80 @@
+  - [ ] `none`
+  - [ ] Other (please elaborate)
+
+# Are you aware that not handling rejections may cause memory leaks or denial of service vulnerabilities?
```
```suggestion
# Do you believe that not handling rejections may cause memory leaks or denial of service vulnerabilities?
```
this is a very leading question; if the third option is actually an answer, then it is not a given that this is the case, which means "are you aware that" is inappropriate, since it implies that this is objectively the case.

mmarchini replied:
Let's wait for @mcollina reply above. We can remove this question if it's considered too leading/biased/argumentative.

benjamingr replied:
I find @ljharb 's suggestions equally leading by the way :]

I recommend we ask something like:

```suggestion
# How do you deal with managing resources wrapped by promises when an unhandled rejection occurs?
```

Or something similar that indicates that the issue is:

 - I have a handle - for example a database connection.
 - the method containing that handle throws (for example releasing the connection).
 - Since it's an async method (if it was sync the server would crash) - I get an unhandled rejection.
 - If I don't quit or explicitly deal with it, it's super easy to create cascading failures and cause denial of service vulnerabilities.

I think it's important that this is not _all_ unhandled rejections but a particular sort, but on the other hand that this is a very real issue.

ljharb replied:
I agree - since it's likely a small number of unhandled rejection cases, that *if* they occur, can cause very serious problems, it's important to avoid leading questions in any direction.

benjamingr commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,80 @@
+
+## How are you handling rejections today?
+
+When consuming Promises, async functions or thenables, which of the options below do you use to handle rejections?
```
Explain what a rejection is here? I'd assume a large number of people don't know what rejections actually are and assume they are just "async errors" that come from async functions.

benjamingr commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,80 @@
+  - [ ] Ignore the rejection
+  - [ ] Not writing Promise-based code or using Promise-based libraries
+
+## Do you use `process.on('unhandledRejection')` as a global handler for unhandled rejections?
```
I'd ask if they know what the unhandledRejection hook is first?

benjamingr commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,80 @@
+  - [ ] Desktop applications
+  - [ ] Other (please elaborate)
+
+## When a rejected promise doesn't have a catch handler, what is the behavior of your application today? 
```
I think we should show an example of this so we're sure people understand what this means.

For example:

```js
async function error() {
  throw new Error();
}
error();
```

benjamingr commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,80 @@
+  - [ ] Exit as soon as possible
+  - [ ] Other (please elaborate)
+
+## What should be the default Node.js behavior for unhandled rejections?
```
Like above, I don't think people understand what unhandled rejections are very well.

I would prefer "when you throw an error from an async function and don't `.catch` it" or something similar that describes that the issue being discussed is errors thrown and not just `Promise.reject` calls. A lot of people don't know that both of these create rejections.

mmarchini commented:
Sorry y'all for the delay updating the survey. Thank you @benjamingr @devsnek and @ljharb for this round of feedback! I think I incorporated everything in this last commit:

  1. Added a preface explaining what rejections and unhandled rejections are (with examples)
  2. Expanded some questions to cover more cases
  3. Removed the last question as I couldn't think of a way to phrase the question + answers without somehow leading the participant. If anyone has suggestions on how to rephrase it, we can re-add it. I think the title suggested by @benjamingr [here](https://github.com/nodejs/TSC/pull/857#discussion_r421433352) would be good, but I'm having trouble phrasing answers for it, maybe this should be an text field question where users can describe what they do?

I think we should start to wrap this up so we can send it out and gather results.

ljharb commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+foo().catch(() => console.error("an error occured"));  // 1. Handled
+
+try {
+  await foo();  // 2. Handled
+} catch(e) {
+  console.error("an error occured")
```
```suggestion
  await foo();
} catch(e) {  // 2. Handled
  console.error("an error occured")
```

danbev commented on surveys/promise-rejections/survey.md:
```diff
@@ -1,5 +1,70 @@
 # Node.js Promise reject use case survey
 
+Today, Node.js handles unhandled rejections by emitting a deprecation warning to stderr. The warning shows the stack where the rejection happened, and states that in future Node.js versions unhandled rejections will result on Node.js exitting with non-zero status code. We intend to remove the deprecation warning, replacing it with a stable behavior which might be different from the one described  ...
```
Nit: 
> in future Node.js versions unhandled rejections will result on Node.js exitting with non-zero status code.

 in future Node.js versions unhandled rejections will result `in` Node.js `exiting` with non-zero status code.

danbev commented on surveys/promise-rejections/survey.md:
```diff
@@ -1,5 +1,70 @@
+
+## What is a Promise rejection?
+
+A Promise rejection indicates that something went wrong while executing a Promise or an async function. Rejections can occur on several situations: thorwing inside an async function or a `Promise` executor/then/catch/finally callback, when calling the `reject` callback of an `executor`, or when calling `Promise.reject`.
```
Nit:
> Rejections can occur on several situations: thorwing inside an async function

Rejections can occur `in` several situations: `throwing` inside an async function

devsnek commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+
+## Are you currently using Promises, async/await, a mix, or neither?
+
+Assume using as creating your own `new Promise`, `Promise.resolve`, and `async function`; or using libraries that produce Promises (that you use with `.then`/`.catch`, or with `.await`)
```
```suggestion
Assume using as creating your own `new Promise`, `Promise.resolve`, and `async function`; or using libraries that produce Promises (that you use with `.then`/`.catch`/`.finally`, or with `await`)
```

mhdawson replied:
"Assume using as creating" still does not make sense to me. Not sure what its trying to communicate.

ljharb commented:
@mcollina the trick tho, is that *anything* can be the cause of leaks, and most unhandled rejections *don't* cause leaks - i don't think it's actually a given that unhandled rejections make leaks that much more likely, so it's tough to word a question that lacks the bias of that assumption.

mcollina commented:
The way our servers APIs are designed they _always_ cause a leak if a chain is broken and a response is not sent. The safe behavior is to crash if the user has not decided this is ok for them, e.g. they prefer to restart every few hours vs every minute.

The key question for me if our users prefers to leak memory vs crashing by default. It is essentially a control question for all the others.

---

An interesting data point is that Deno crashes with unhandled rejections.

devsnek commented:
"did you know that if your code looks like `(req, res) => { f().then((r) => res.send(r)); }` you might be leaking http connections"

ljharb commented:
@mcollina tbh that sounds like a design flaw in your servers more than any inherent property of unhandled rejections. I'm all for trying to gauge users' familiarity with memory leaks, but it doesn't feel appropriate to me to imply something universal that's only true as a result of specific coding patterns.

mcollina commented:
I agree this is a design flaw of Node.js, however we are not at a luxury of changing it without massive ecosystem breakage as we would need to redesign our entire http (and net, and tls). This survey is about Node.js behavior, so I think it’s totally relevant.

puzpuzpuz commented:
Unhandled rejections may leave (and in most cases, they do) applications in non-deterministic state, whether it's internal in-memory state (incl. mem leaks), OS/external resources (handles, DB connections, etc.), or external state (consistency of data in DBs/caches, etc.).

It would be great if we would add this or a similar text into `## What is a Promise rejection?` section, so that users are aware of potential problems caused by unhandled rejections.

> @mcollina tbh that sounds like a design flaw in your servers more than any inherent property of unhandled rejections.

In general, I'd say that it's a design flaw of promises themselves, as they make it so easy to swallow an exception instead of forcing developers to handle rejections explicitly.

mcollina commented:
I'm good with including such a text.

puzpuzpuz commented on surveys/promise-rejections/survey.md:
````diff
@@ -0,0 +1,152 @@
+```
+
+An unhandled rejection is a rejection which hasn't been handled yet. It might be handled in the future, like example 4, but it might also stay unhandled forever (like example 3).
+
````
Suggestion. Add the following paragraph, so that users are aware of potential problems caused by unhandled rejections.:
```
Unhandled rejections may leave your application in a non-deterministic state, whether it's internal application state (including memory leaks), external resources used by your application (say, file handles or database connections), or external state (say, consistency of data in a database).
```

I don't mind changing the text in a more readable/concise manner.

benjamingr replied:
I think that's a bit scary and Jordan and Gus would probably object to that phrasing. I suggest alternatively:

> Certain unhandled rejections may in rare cases leave your application in a non-deterministic and unsafe state, whether it's internal application state (including memory leaks), external resources used by your application (say, file handles or database connections), or external state (say, consistency of data in a database).

I think it's important to communicate that it's not normal for any unhandled rejection to leave your app in inconsistent state but that it's possible.

puzpuzpuz replied:
I'm fine with the proposed change.

benjamingr commented on surveys/promise-rejections/survey.md:
````diff
@@ -0,0 +1,152 @@
+
+```js
+
+Promise.reject()  // This will result in a rejection
````
I mean, I know it's a bit silly but I would prefer if all these examples show `.reject(new Error())` just to not show people code that rejects with undefined without a stack trace 😅

benjamingr commented on surveys/promise-rejections/survey.md:
````diff
@@ -0,0 +1,152 @@
+    a();  // This will result in a rejection
+}
+```
+
````
I would add a line or comment here explaining that just adding `async` before a function can transition an exception from an uncaught exception to an unhandled rejection.

benjamingr commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+
+## What is an unhandled rejection?
+
+Rejections are designed to be handled asynchronously. In JavaScript, Promises can be handled when awaiting for a promise inside a try/catch block, or via .catch calls:
```
I don't think that's true, rejections are _designed_ to be handled asynchronously so that errors are always in the same timing and users don't have to do `.catch` and `} catch {` both.

Promise are also generally not "handled" they are just "reacted to".

So I would either remove this line or explain that while _typically_ rejections are handled synchronously in JavaScript it is _possible_ to handle them asynchronously by attaching a `.catch` later on (inside a setTimeout) for example.

mmarchini replied:
> Promise are also generally not "handled" they are just "reacted to".

Even if "reacted to" is more correct from a spec perspective, I think it's important to consider the audience we're targeting, who might not be familiar with the term "reacted to" (for example, [on the MDN page for Promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise), the terminology "reacted to" is never used, but handlers is. Also, using "reacted to" might cause dissonance with "unhandled rejection" in the survey. If others feel strong about this terminology we can change, but I think sticking to "handled" here will be clearer for participants.

I see your point about the general phrasing though. This is a hard topic to phrase :). For example, IMO `_typically_ rejections are handled synchronously` gives the wrong impression that you can react to a rejection on the same turn it happened, when in truth it will only be handled in the future, even if we do `try { async Promise.reject() } catch { }`. I'll try to rephrase it in a way it covers what you're saying.

benjamingr commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+
+foo(); // 3. Unhandled, but execution continues
+
+const rejected = foo(); // 4. Unhandled on current event loop tick
```
I would prefer "task" or something other than "event loop tick" which is a bit vague in terms of terminology.

ljharb replied:
Hmm, I find “event loop tick” precise and “task” to me means “promise-returning function” :-)

benjamingr replied:
I'm honestly not sure what "event loop tick" means, I've never thought of it as a loop and I doubt users care about [internals](https://nodejs.org/fa/docs/guides/event-loop-timers-and-nexttick/) (even though that article isn't exactly accurate it's the best one I know).

If you think it's clearer I will concede, it's possible the vague terminology is a lot clearer than task/microtask.

ljharb replied:
stuff i send to newcomers in irc include https://nodesource.com/blog/understanding-the-nodejs-event-loop , https://www.youtube.com/watch?v=8aGhZQkoFbQ , https://www.youtube.com/watch?v=cCOL7MC4Pl0 - "event loop" is generally a well known concept from the browser. i've heard both "tick" and "turn" to describe an iteration of the event loop.

mmarchini replied:
I rarely (if every) hear Node.js developers refer to it as `task`. It's worth noting that this survey is targeted to Node.js developers, who may not be familiar with ECMAScript spec terminology.

I like `turn` better than `tick`, thanks @ljharb, I'll update this comment.

benjamingr commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+
+When consuming Promises, async functions or thenables, which of the options below do you use to handle rejections?
+
+(Multiple choice)
```
I mean, 'ignore the rejection' isn't really fair IMO because it implies the rejection is "ignored" where propagating rejections is what people do in 99% of cases I'd assume (because of how exceptions work).

So in most cases I'd assume he answer would be "propagate the error to my caller" since most code is not wrapped in try/catch.

mmarchini replied:
I'll replace with "Leave the handling to my caller", as "propagate" is a commonly used term when throwing. If others prefer "propagate" anyway, we can use it instead though.

benjamingr commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+  - [ ] Ignore the rejection
+  - [ ] Not writing Promise-based code or using Promise-based libraries
+
+## Do you know that Node.js has a global handled for unhandled rejections (`process.on('unhandledRejection')`)? If so, do you use it?
```
Can we please add (for comparison) the same question about `uncaughtException`? I want to know if the people answering aren't aware of the promises hook are also not aware of the non-promises hook.

jasnell replied:
```suggestion
## Do you know that Node.js has a global handler for unhandled rejections (`process.on('unhandledRejection')`)? If so, do you use it?
```

benjamingr commented on surveys/promise-rejections/survey.md:
````diff
@@ -0,0 +1,152 @@
+error();
+```
+
+  - [ ] Node.js logs a warning alongside a deprecation notice, execution continues
````
I would note here "this is the default Node.js behavior today".

BridgeAR replied:
I would keep it as is. It should not be an important factor what the current default is, right?

benjamingr replied:
I'm not sure, there is a trade off. I only meant mentioning this here (and below) to better explain what it means.

If you believe it will create bias I will concede.

mmarchini replied:
I'll add it here and below (as suggested by @benjamingr), but we can remove if folks feel strong about it creating bias.

benjamingr commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+  - [ ] Node.js logs a warning alongside a deprecation notice, execution continues
+  - [ ] Logs a warning, continue running, no deprecation warning
+  - [ ] Logs a warning, continue running, no deprecation warning, exits with an error code when program finishes
+  - [ ] Exit as soon as possible
```
I would note here "this is the default Node.js behavior for uncaught exceptions today".

BridgeAR replied:
I suggest to change it to: `Raises an exception that ends the application`.

mmarchini replied:
> I suggest to change it to: Raises an exception that ends the application.

Is that accurate though? `process.on("unhandledRejection", process.abort)` will not raise an exception, for example.

benjamingr commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+
+Consider the following modes:
+
+  - `strict`: throw and crash as soon as possible (on nextTick). Can't be overriden by unhandledRejection listener
```
"This is similar to what Node.js does for regular uncaught exceptions"

benjamingr commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+
+Consider the following modes:
+
+  - `strict`: throw and crash as soon as possible (on nextTick). Can't be overriden by unhandledRejection listener
```
"This is similar to what Node.js does when you pass --abort-on-uncaught-exception" ?

BridgeAR replied:
This is not absolutely correct. It does not crash but it raises an uncaught exception by default.

Maybe change it to:

```  - `strict`: raise an uncaught exception (on nextTick) similar to `throw new Error()` that is not caught. `unhandledRejection` listeners do not prevent raising the exception```

benjamingr commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+Consider the following modes:
+
+  - `strict`: throw and crash as soon as possible (on nextTick). Can't be overriden by unhandledRejection listener
+  - `throw`: throw and crash as soon as possible (on nextTick). Can be overriden by unhandledRejection listener
```
"This is similar to what Node.js does for regular uncaught exceptions"?

BridgeAR replied:
Maybe change it to:

```- `throw`: raise an uncaught exception (on nextTick) similar to `throw new Error()` that is not caught. `unhandledRejection` listeners take precedence and prevent raising the exception```

benjamingr commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+
+  - `strict`: throw and crash as soon as possible (on nextTick). Can't be overriden by unhandledRejection listener
+  - `throw`: throw and crash as soon as possible (on nextTick). Can be overriden by unhandledRejection listener
+  - `warn`: outputs a warning as soon as possible (on nextTick). Continues running after the warning is emitted. If the process exits and no status code was set, the process exits with a success code
```
"This is similar to what browser consoles do"?

BridgeAR replied:
What about explaining that all of this is happening on nextTick above in a note and to remove the (on nextTick) from the entries here?

benjamingr commented:
> Removed the last question as I couldn't think of a way to phrase the question + answers without somehow leading the participant. 

## How do you deal with managing resources wrapped by promises when an unhandled rejection occurs?

Node.js code may wrap resource managing code in async/await:

```js
// if this method wasn't async, node would crash by default
myEmitter.on('event', async () =>{
   // react to event
  databaseConnection.release(); // throws an error
});
```

If the above code throws, a database handle might leak which can in some cases eventually lead to a cascading failure and a denial of service.

How do tools or servers you author deal with this case?

 - [ ] I don't author code where this might be an issue (for example, code I author does not connect to third party resources).
 - [ ] I take extra care to deal with these cases individually and perform monitoring on database handles with alerts.
 - [ ] I make sure to restart my server if code like `databaseConnection.release` a throws, like with `uncaughtException`s.
 - [ ] In theory this can be an issue with code I author but in practice things have been working out fine for me and I ignore these errors.
 - [ ] The server keeps running, I log uncaught exceptions and unhandled rejections to monitoring and use a tool to notify engineers if such a bug occurs.

And probably another question indicating what Node.js default should be _in that case_.

jasnell commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+  - [ ] I don't use `--unhandled-rejections`
+  - [ ] I didn't know `--unhandled-rejections` existed
+
+## Are you using Promises in customer facing, production applications?
```
'customer facing' here is problematic... every line of code has a customer, it just varies who that customer is. I would restructure this to ask something like:

```
Are you using Promises in any of the following kinds of applications? (check all that apply)

[ ] Production-level code
[ ] Tests
[ ] Benchmarks
[ ] Build pipeline / infrastructure
[ ] Examples and Demos
[ ] Other ____
```

BridgeAR commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+  - [ ] I use `process.on('unhandledRejection')`
+  - [ ] I don't use `process.on('unhandledRejection')`
+  - [ ] I didn't know `process.on('unhandledRejection')` existed
+
```
I suggest to add another question or entry above that mentions libraries such as `https://github.com/mcollina/make-promises-safe`.

benjamingr replied:
```
 - [ ] I use a third-party library (like make-promises-safe) to deal with unhandled rejections.
```

Or something similar sounds good to me.

BridgeAR commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+  - [ ] Desktop applications
+  - [ ] Other (please elaborate)
+
+## When a rejected promise doesn't have a catch handler, what is the behavior of your application today? 
```
Maybe add a note that asks for the biggest projects / most commonly worked on code? Most people won't only have a single application.

BridgeAR commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+  - [ ] Logs a warning, continue running, no deprecation warning
+  - [ ] Logs a warning, continue running, no deprecation warning, exits with an error code when program finishes
+  - [ ] Exit as soon as possible
+  - [ ] Other (please elaborate)
```
We might want to add more possible situations here. Suggestions welcome :-)

benjamingr replied:
I mean, I'd suggest GC again (`[ ] Warn as soon as possible and exit as soon as it can prove it is impossible to add a catch handler to the promise.`) but I'm afraid for my safety after how much backlash that suggestion got last time 😅

mmarchini replied:
Is that something that can be implemented in userland? Otherwise it wouldn't fit the question.

BridgeAR commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,152 @@
+  - [ ] Exit as soon as possible
+  - [ ] Other (please elaborate)
+
+## What should be the default Node.js behavior for unhandled rejections?
```
We might want to add a note where `unhandled rejections` are described clearly to understand what they stand for?

mmarchini replied:
I think the introduction above covers that? I'm open to suggestions on how to improve it though.

mmarchini commented:
> I would like the last question to be added again, and I was surprised it was removed without cc me (I requested it to be added).

@mcollina sorry, my intention was not to leave the question out, I removed it because it needed to be rephrased. Also, I didn't cc you before removing because I thought the cc [here](https://github.com/nodejs/TSC/pull/857#discussion_r421021691) (where we here discussing the question) was enough, but in retrospect I should've cc'd you again before removing it. I'll re-add the question following the suggestion on everyone's comments.

> And probably another question indicating what Node.js default should be _in that case_.

I'm not sure about that, adding another question (especially phrasing as "in that case") seems to be leading/incentivizing the participant to change their response. What others think?

I'll read everyone else's comments and update the PR later today.

BridgeAR commented:
@mmarchini thank you very much for taking the initiative and for driving this forward!

mmarchini commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,211 @@
+
+## What is an unhandled rejection?
+
+There are two ways to handle rejections: by attaching a `.catch` handler to it,
+or by awaiting on the promise within a try/catch block. In both cases, the
+handling of the rejection (the execution of the callback passed to `.catch`, or
+the execution of the `catch {}` block) will happen in a future turn of the
+event loop.
+
+Promises are designed so that attaching handlers or awaiting can be done at any
+point in time, from when the Promise was created (possibly while it's still
+pending), to right before the program finishes execution.
+
+A rejection is considered unhandled from the point it happens until the point
+where a handler is attached to the Promise or the Promise is awaited within a
+`catch {}` block. Below are a few examples of handled and unhandled rejections.
```
@benjamingr what do you think?

benjamingr replied:
There are a couple of technical inaccuracies here:

 - There are a lot more ways to handle rejections (put a promise in a Promise.all or other combinator, `resolve` another promise with it in the promise constructor etc).
 - It is not clear what "in a future turn of the event loop" means in the context of handling of the rejection but that's usually not the case - since it happens after nextTicks run but before any "event loop stuff" like timers and I/O.

But I think it's fine for us to be not 100% technically accurate when explaining terms and honestly we could keep talking about this for years :D 

I think overall the survey is LGTM and I'm +1 on sending it.

mmarchini commented:
Wow, so many good suggestions, thanks y'all! I think I addressed everything (left some comments on any proposed changes I didn't apply yet). If I missed anything, please let me know.

benjamingr commented:
Latest revision LGTM, thank you for taking the time to do all this work :)

ljharb commented on surveys/promise-rejections/survey.md:
````diff
@@ -0,0 +1,211 @@
+
+```js
+// if this method wasn't async, node would crash by default
+myEmitter.on('event', async () =>{
````
```suggestion
myEmitter.on('event', async () => {
```

mcollina commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,211 @@
+
+As we can see in the examples, an unhandled rejection might be handled in the future, like example 4, but it might also stay unhandled forever (like example 5).
+
+Certain unhandled rejections may in rare cases leave your application in a non-deterministic and unsafe state, whether it's internal application state (including memory leaks), external resources used by your application (say, file handles or database connections), or external state (say, consistency of data in a database).
```
Can you add:

> As an example, the following server is not sending a response back to the client, causing a socket leak and a possible Denial of Service attack:

```js
const http = require('http')
const server = http.createServer(handle)

server.listen(3000)

function handle (req, res) {
  doStuff()
    .then((body) => {
      res.end(body)
    })
}

function doStuff () {
  if (Math.random() < 0.5) {
    return Promise.reject(new Error('kaboom'))
  }

  return Promise.resolve('hello world')
}
```

```

mmarchini replied:
Done!

mmarchini commented:
@mcollina I believe your [requested changes](https://github.com/nodejs/TSC/pull/857#pullrequestreview-411478595) were addressed?

dfabulich commented on surveys/promise-rejections/survey.md:
````diff
@@ -0,0 +1,235 @@
+}
+```
+
+Adding the `async` keyword to a function will turn any exceptions thrown inside that function (or any throw propagated from other functions called within it) into a rejection. The same happens when refactoring callback based code that throws into async functions / Promises. Below is an example of a callback based code refactored to Promises where exceptions become rejections:
````
This paragraph says that the example is below, but the example appears to be missing…?

mmarchini replied:
I think something went wrong during a rebase, the example was indeed supposed to be below, but I can't find the example anymore :facepalm:

I'll rewrite the example.

dfabulich commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,235 @@
+
+## Do you know that Node.js has a global handler for uncaught exception (`process.on('uncaughtException')`)? If so, do you use it?
+ 
+  - [ ] I use `process.on('unhandledRejection')`
```
These should say `uncaughtException` … the answers are incorrectly copied and pasted from an earlier question.

dfabulich commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,235 @@
+  - `warn-with-error`: outputs a warning as soon as possible. Continues running after the warning is emitted. If the process exits and no status code was set, the process exits with an error code
+  - `none`: do nothing
+
+For all the modes, the action (raise an exceptiom output a warninig) will happen on `nextTick`.
```
typo, should be "warning"

dfabulich commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,235 @@
+
+const rejected = foo(); // 5. Unhandled on current event loop turn
+
+// 4. ... but handled in a future turn of the event loop when the setTimeout
```
The numbering here is inconsistent... there are two points numbered 4, and the "example 4" text below it is confusing.

mmarchini replied:
Oops, I messed up the numbering here. I'll fix it.

> and the "example 4" text below it is confusing

Any suggestions on how to make it more approachable?

dfabulich replied:
I just meant that you can't say "example 4" when there are two examples 4 ;-)

mmarchini replied:
Oh, I see. So the idea is that the "second" example 4 is a continuation of the first (it's still the same example, but it shows how a catch can be attached at a later point in time). I'll see if there's a way to rephrase and reorganize so it keeps the message but in a clear way (or, if you have any suggestions, I'm open to hear :) )

ljharb commented on surveys/promise-rejections/survey.md:
````diff
@@ -88,6 +88,30 @@ As we can see in the examples, an unhandled rejection might be handled in the fu
+As an example, the following server is not sending a response back to the client, causing a socket leak and a possible Denial of Service attack:
+
+```js
+const http = require('http')
+const server = http.createServer(handle)
+
+server.listen(3000)
+
+function handle (req, res) {
+  doStuff()
+    .then((body) => {
+      res.end(body)
+    })
+}
+
+function doStuff () {
+  if (Math.random() < 0.5) {
+    return Promise.reject(new Error('kaboom'))
+  }
+
+  return Promise.resolve('hello world')
+}
````
total nit, but this way it'll match the most common style in the ecosystem
```suggestion
const http = require('http');
const server = http.createServer(handle);

server.listen(3000);

function handle(req, res) {
  doStuff()
    .then((body) => {
      res.end(body);
    });
}

function doStuff() {
  if (Math.random() < 0.5) {
    return Promise.reject(new Error('kaboom'));
  }

  return Promise.resolve('hello world');
}
```

mmarchini replied:
~oh no, we'll start discussing semicolons now~

mmarchini replied:
I should apply a linter here or something, I just noticed there's no consistency across the examples :/

mmarchini replied:
Ran eslint on code blocks, should be more following a single style now

Raynos commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,235 @@
+  - [ ] Native Promises (`new Promise`, `Promise.resolve`, `Promise.reject`)
+  - [ ] Third-party Promises (for example: Bluebird)
+  - [ ] async functions
+  - [ ] Not writing Promise-based code or using Promise-based libraries
```
Do we want to add

```
 - [ ] Add `process.on('unhandledRejection', handleError)` and handle errors in a single place.
```

I've written code for Electron / Lambda where I ignore promises rejections and handle them globally.

mmarchini replied:
I think you meant to comment in the question below? Either way, this is covered in another question (for this case, I'd expect participants to answer "leaving handling to my caller")

Raynos replied:
Oh your right, i did leave my comment on the wrong block. switching between text and rendered HTML fail.

Raynos replied:
Left the comment in the correct place. oops, expanded upon it. this one can be resolved. :see_no_evil:

Raynos commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,235 @@
+// if this method wasn't async, node would crash by default
+myEmitter.on('event', async () => {
+   // react to event
+  databaseConnection.release(); // throws an error
```
I don't understand the comment about throwing an error. If the method to "release" the database connection throws then either A) It released anyway or B) its impossible to release the connection

Maybe you meant a different example where an error is thrown before `release()` and the `release()` method is never called.

```js
await databaseConnection.getValue() // throws an error
databaseConnection.release()
```

mmarchini replied:
@benjamingr suggested this example, so I'll let him answer.

benjamingr replied:
> A) It released anyway or B) its impossible to release the connection

Either releasing the connection failed (in which way you probably want to restart the server) or there exists a more "severe" release API for example:

```
database.release(); // EREAD_IN_PROGRESS
database.forceRelease(); // works
```

I think it's also fine for the example to show release never being called (by throwing an error earlier) and it might simplify things.

Raynos replied:
Adding `// throws an error; oops forgot to call database.forceRelease()`

Would resolve my confusion.

mmarchini replied:
Let's go with the getValue example above ("release never called"), I think it simplifies and clarifies things.

mmarchini commented:
I'll update the survey Wednesday with some of the latest suggestions. I intend to land it before the next TSC meeting unless there's an outstanding issue.

Raynos commented on surveys/promise-rejections/survey.md:
```diff
@@ -0,0 +1,235 @@
+
+  - [ ] `.catch()`
+  - [ ] `try / catch` wrapping an `await` operation
+  - [ ] Leave the handling to my caller
```
Do we want to consider an extra option

```
 - [ ] Add `process.on('unhandledRejection', handleError)` and handle errors in a single place.
```

This is different then `Leave the handling to my caller` because you can then write code where you do

```js
async function doSomething() {}`

doSomething() // no await / catch;
// or more realistically `x.on('someevent', doSomething)`
```

The caller does not have `.catch()` nor `try / catch` ;

This is mostly about using `async / await` without `try / catch / .catch()` using a library like https://github.com/scopsy/await-to-js

mmarchini replied:
I still think this is covered in the "leave the handling to the caller" (maybe changing to "leave the handling to someone else"?) and (especially) by the other question. Adding an option for unhandledRejection handler here would mix localized handling with global handling, and IMO that's not the goal of this question.

I would like to hear others opinion on this though.

mmarchini replied:
Maybe this question shouldn't even be in the survey (sometimes I use .catch, try catch and "leave to others" all in the same code).

mmarchini replied:
Kept the question and expanded the "leave handling..." answer.

mmarchini commented:
Landed in 83dc49611e9c

nylen commented:
Who does this survey get sent out to? I'd like to participate.

nylen commented:
Also, the fact that a survey is being done at all seems to contradict what the Node.js command-line output says:

```
(node:12314) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code**

Sounds like the decision was already made some time ago?

dfabulich commented:
> Sounds like the decision was already made some time ago?

As of today, TSC members are in disagreement about it, regardless of what the deprecation message said when it was written. We're hoping that this survey will help TSC to reach consensus.

mmarchini commented:
@dfabulich TSC members being in disagreement is not a problem, since conflict is resolved via vote. But we want to hear the community since this is a sensitive topic.

dfabulich commented:
It's may be too late, but is there any chance we could call it `warn-with-error-code`? This name emerged while reviewing https://github.com/nodejs/node/pull/33475 and I think it's a little clearer. 

(It's OK if the survey goes out as-is, but it would definitely be nicer to tweak this one last thing.)

mmarchini commented:
For the purpose of the survey I don't think the name matters that much, since we explain what each of the modes are (and we listed modes which are not part of Node.js yet). But there's still time to change it, yes.

nylen commented:
@mmarchini has this survey started yet?

mmarchini commented:
No. I forwarded the request to the Foundation, but they don't have an ETA yet. Either way, this is a breaking change, therefore even if it lands on master soon it won't be included in any release before we release v15.0.0 in October, so we have time to run the survey and reach a decision.

dfabulich commented:
When the survey goes out, should I expect to see a comment here? I just want to make sure I don't miss the announcement.

mmarchini commented:
Yes, I'll share a comment here and on the pull requests and issues.

dfabulich commented:
It's been another whole month. I think it might be time to pester someone again?

mmarchini commented:
I'm in contact with them on the OpenJS foundation slack channel. Let me ping them again to see how the progress is going,

mmarchini commented:
The survey is available at: https://www.surveymonkey.com/r/FTJM7YD

We also wrote an accompanying blog post for extra context: https://medium.com/@nodejs/node-js-promise-reject-use-case-survey-98e3328340c9

The survey will run for at least two weeks, at which point we'll evaluate if the number of replies is enough for us to move forward, otherwise we might extend it for a week or two. Please fill out the survey as it will help us decide the future of unhandled promise rejections on Node.js!

nylen commented:
Thanks for linking to the survey. There is no place for additional comments apart from the "Other" field so I will leave them here.

Please take the following into account regardless of the decision ultimately (re)made by the TSC:

- Unhandled rejections **must** have clean, clear stack traces with no extraneous output: https://github.com/nodejs/node/issues/17871
- It **must** be possible to control the default behavior of unhandled rejections using a single line of **application code** (`--unhandled-rejections` is specifically not adequate because Node.js process launches are not always under control of the application developer): https://github.com/nodejs/node/issues/33577

benjamingr commented:
@nylen note: I think that message comes off as needlessly aggressive. By making the word "must" bold you are not doing yourself a service.

Instead: 

 - If you believe that it must be possible to control the behavior of unhandled rejection with a single line of application code (and `process.on('unhandledRejection', e => { throw e })` is insufficient) I recommend you convince us (not as in me or Mary - as in consensus on that issue) - I am personally not convinced in https://github.com/nodejs/node/issues/33577 .
 - Regarding stack traces - didn't Ruben fix that? (in https://github.com/nodejs/node/issues/17871 )

I am thankful for your involvement and engagement in this space (we need more people who understand this problem domain involved). Please if at all possible keep discussions friendly (honestly it's very silly to not have friendly discussions and getting along given how few people are involved and how historically most of the problems were solved by discussions). We did have quite a few people walk away from the problem space in the past because of the way conversations were held.
